### PR TITLE
[WIP] BF : Handle small sphere radius

### LINKED ISSUE #873: NiftiSpheresMasker crashes when radius is small
When trying this:

```python
from nilearn.datasets import fetch_adhd
from nilearn.input_data import NiftiSpheresMasker

dataset = fetch_adhd(n_subjects=1)
dmn_coords = [(0, -52, 18), (-46, -68, 32), (46, -68, 32), (0, 50, -5)]
masker = NiftiSpheresMasker(seeds=dmn_coords, radius=1)
data = masker.fit_transform(dataset.func[0])
```

The spheres are then empty.

### PR DESCRIPTION
**Pull request by mrahim**
Fix #873

### CODE REVIEW

File: nilearn/input_data/nifti_spheres_masker.py [hunk 1]
```diff
@@ -50,6 +50,8 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
 
     clf = neighbors.NearestNeighbors(radius=radius)
     A = clf.fit(mask_coords).radius_neighbors_graph(seeds)
+    if np.any(A.sum(axis=1) == 0):
+        A = clf.fit(mask_coords).kneighbors_graph(seeds, 1)
```
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_spheres_masker.py]: Kind of logic of what we should do
```python
     for i, seed in enumerate(seeds):
         try:
             if A[i].sum() == 0:
                 A[i] = clf.kneighbors_graph(seeds, 1)
         except ValueError:
             # seed is not in the mask
             pass
```

File: nilearn/input_data/nifti_spheres_masker.py [hunk 2]
```diff
@@ -63,7 +65,6 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
     if not allow_overlap:
         if np.any(A.sum(axis=0) >= 2):
             raise ValueError('Overlap detected between spheres')
-
```
Comment (banilo) [inline nilearn/input_data/nifti_spheres_masker.py]: = merge conflict generator?

**Comment by GaelVaroquaux** [comment]
@mrahim : any update on this? We'd love to include this on the upcoming 0.2.2 release.

**Comment by mrahim** [comment]
Yes, reviews are welcome.

File: nilearn/input_data/nifti_spheres_masker.py [hunk 3]
```diff
@@ -49,8 +49,21 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
...
     clf = neighbors.NearestNeighbors(radius=radius)
-    A = clf.fit(mask_coords).radius_neighbors_graph(seeds)
-    A = A.tolil()
+    clf.fit(mask_coords)
+    A = []
+    for i, s in enumerate(seeds):
+        # Check seed is inside the mask
+        dist, _ = clf.kneighbors(s, 1)
+        if dist > 2:
+            raise ValueError('Seed #%i is too far' % i)
```
Comment (bthirion) [inline nilearn/input_data/nifti_spheres_masker.py]: Sorry, I don't see clearly why you chose 2 as a threshold. It is a distance  in mm ? What if the voxel grid is coarse ?

File: nilearn/input_data/nifti_spheres_masker.py [hunk 4]
```diff
@@ -49,8 +49,21 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
...
+        # Select voxels inside radius
+        a = clf.radius_neighbors_graph(s)
+        # If Radius is too small
```
Comment (bthirion) [inline nilearn/input_data/nifti_spheres_masker.py]: I am also missing something here: how is the radius defined ?

**Comment by bthirion** [comment]
Please add a test.

File: nilearn/input_data/nifti_spheres_masker.py (hunk 3 above)
Comment (mrahim) [inline nilearn/input_data/nifti_spheres_masker.py]: Yes, I don't know how to detrmine a too far seed.  Any suggestions?

File: nilearn/input_data/nifti_spheres_masker.py (hunk 4 above)
Comment (mrahim) [inline nilearn/input_data/nifti_spheres_masker.py]: The radius is the input argument of the sphere (in mm).

**Comment by lesteve** [comment]
Closing in favour of #946.
